### DISCUSSION TITLE
question on an implementation detail

### DISCUSSION THREAD

**Discussion by vonovak** [tamagui/tamagui#1690] [Q&A]
Hello, firstly, thanks for tamagui and for takeout, which I'm using. I'd like to ask a question:

1) consider that this is rendered a 50 times in a screen in a android / ios RN app

```
<XStack
    backgroundColor="$color"
>
    ...
</XStack>
```

I'm assuming this gets rendered as a `<View style={{ backgroundColor: xyz }} />` down the line (?). How many instances of `{ backgroundColor: xyz }` (the JS object) are created in the memory?

Is it possible / would you guys be open to using `DynamicColorIOS` on iOS to control colors for dark and light themes?

Thank you!

**Comment by natew** [comment]
This is interesting, we could optimize to this using the optimizing compiler and internally in tamagui as well. Didn't know this existed, wish is was a more generic API but I assume once they support Android they will do that. Thanks for the heads up, @TheEhsanSarshar can help make this work with the optimizing compiler and I can make it work in core.

**Reply by vonovak** [reply]
@natew thanks for answering. My question was really 2 questions, one about the number of objects in memory - 1 object or 50?
The other was about `DynamicColorIOS`. I'm guessing you answered the second question. Can you answer the first as well? Thank you.

With regard to `DynamicColorIOS`, I don't think there will be an Android support for this because switching colors at runtime isn't really possible so easily. One would have to re-render the views, AFAIK. There is also [PlatformColor](https://reactnative.dev/docs/platformcolor) that works across platforms but that one just "taps into" the colors of the platform.

**Comment by byteab** [comment]
@vonovak If backgroundColor is something static ( not from the theme or a runtime value ), then the compiler for react native will extract it and use `Stylesheet.create()`, which means you only have one instance of that style doesn't matter how manytimes you re-render.
for dynamic styles ( from theme or rutime ) we can use `useMemo` I'll work on that this weekend.
as a workaround you can use this if you have some critical components that need ultimate optimization
```
const style = useMemo(() => ({
     backgroundColor: theme.background.get()
}), [theme])

<Stack width={20} height={40} style={style} />
```

**Reply by vonovak** [reply]
thanks! yeah, I was referring to `backgroundColor="$color"` where `$color` is a color token that comes from tamagui config. ~So one instance only, thank you! :)~
edit: it comes from theme so as I understand it'll not be optimized. Okay...

**Reply by vonovak** [reply]
@TheEhsanSarshar It's not so much about re-rendering, my point was about avoiding having the same object in memory many many times.

would you mind pinging me on the PR that does this style optimization, when ready? Anyways, I just wanted to know if the compiler does some optimizations here or not. Thank you! 🙂

**Reply by byteab** [reply]
yeah when you extract them, it means one object for any number of components or re-renders. it basically uses react native stylesheet API.
```js
const Foo = () => {
     return <View style={styles.view} />
} 
const styles = stylesheet.create({
    view: {
       backgroundColor: 'red'
    }
})
```
to clear up one thing here, there's no way to use the same dynamic style object for any number of the same component in react native unless you create a style in the parent component and pass it as a prop to all those 50 instances. 
I'll let you know when that PR is ready!

**Reply by byteab** [reply]
@vonovak 
here is a PR for that
https://github.com/tamagui/tamagui/pull/1757
it is working, I need to take @natew feedback and write some tests.
right now it supports the extraction of theme values which was the hard part.
adding support for dynamic values like using a value from prop or some apis isn't hard to add. but will try to add that in a separate PR. probably this weekend
